### PR TITLE
feat(bpf): Send events via Ring Buffer

### PR DESCRIPTION
**Pull request by dave-tucker**
This uses a ring buffer to send events from eBPF back to userspace.
Doing so allows for our eBPF probes to complete quickly, and pushes
all delta calculation and summary back into userspace.

Moves the logic that resolves comm from procfs into its own pkg.
In addition, a cache is added to avoid hitting procfs every time
we update process metrics.

Removes the bpftest pkg and includes the test programs in the
main kepler.bpf.c file. This prevents drift in the go generate
flags and generally makes it easier to write tests.

Decreases probe execution time by 40% and uses a lot less CPU cycles/instructions per probe invocation.
Also removes a lot of eBPF maps reducing the overall memory required.

Fixes: #1611, #1607, #1354, #1640

### CODE REVIEW

File: pkg/bpf/exporter.go [hunk 1]
```diff
@@ -192,38 +193,221 @@ func (e *exporter) Detach() {
...
+
+func (e *exporter) CollectProcesses() ([]ProcessMetrics, error) {
+	e.mu.Lock()
+	defer e.mu.Unlock()
+
+	metrics := make([]ProcessMetrics, len(e.processMetrics))
+	for pid, m := range e.processMetrics {
+		if pid == 2019461 {
+			fmt.Printf("pid: %d, runtime: %d, cpu_cycles: %d, cpu_instructions: %d\n", pid, m.CPUTime.Total/1000, m.CPUCyles.Total, m.CPUInstructions.Total)
+		}
```
Comment (sthaha) [inline pkg/bpf/exporter.go:395]: debug code?

File: pkg/bpf/exporter.go [hunk 2]
```diff
@@ -192,38 +193,221 @@ func (e *exporter) Detach() {
...
+					return nil
+				}
+				if errors.Is(err, ringbuf.ErrFlushed) {
+					record.RawSample = record.RawSample[:0]
+				}
+				klog.Errorf("reading from reader: %s", err)
+				continue
+			}
+
+			if err := binary.Read(bytes.NewBuffer(record.RawSample), binary.LittleEndian, &event); err != nil {
```
Comment (sthaha) [inline pkg/bpf/exporter.go:237]: shouldn't we use the host.ByteOrder instead?

File: pkg/bpf/exporter.go [hunk 3]
```diff
@@ -192,38 +193,221 @@ func (e *exporter) Detach() {
...
+func (p *PerCPUCounter) Stop(cpu, taskID uint32, value uint64) error {
+	key := uint64(cpu)<<32 | uint64(taskID)
+
+	if _, ok := p.Values[key]; !ok {
+		return nil
+	}
+
+	if value == 0 {
+		return nil
+	}
```
Comment (sthaha) [inline pkg/bpf/exporter.go:311]:
```suggestion
	if value == 0 {
		return nil
	}

```

Comment (sthaha) [inline pkg/bpf/exporter.go:311]:
```suggestion
	if value == 0 {
		return nil
	}
	
	key := uint64(cpu)<<32 | uint64(taskID)
	if _, ok := p.Values[key]; !ok {
		return nil
	}

```

Lets avoid a lookup when possible

File: pkg/bpf/exporter.go [hunk 4]
```diff
@@ -192,38 +193,221 @@ func (e *exporter) Detach() {
...
+
+	var delta uint64
+
+	// Probably a clock issue where the recorded on-CPU event had a
+	// timestamp later than the recorded off-CPU event, or vice versa.
+	if value > p.Values[key] {
+		delta = value - p.Values[key]
+	} else {
+		delta = 0
+	}
```
Comment (sthaha) [inline pkg/bpf/exporter.go:321]:
```suggestion
	var delta uint64 = 0   

	// Probably a clock issue where the recorded on-CPU event had a
	// timestamp later than the recorded off-CPU event, or vice versa.
	if value > p.Values[key] {
		delta = value - p.Values[key]
	}
```

File: pkg/bpf/exporter.go [hunk 5]
```diff
@@ -192,38 +193,221 @@ func (e *exporter) Detach() {
...
+func (e *exporter) handleEvent(event keplerEvent) {
+	e.mu.Lock()
+	defer e.mu.Unlock()
+
+	var p *bpfMetrics
+
+	if _, ok := e.processMetrics[event.Pid]; !ok {
+		e.processMetrics[event.Pid] = newBpfMetrics()
+	}
+	p = e.processMetrics[event.Pid]
```
Comment (sthaha) [inline pkg/bpf/exporter.go:342]: We can extract better performance by locking only read and write to data. 
Have you considered a RWMutex? 

Is the purpose of Mutex to lock the entirely of the exporter or only the processMetrics map?

File: pkg/bpf/exporter.go [hunk 6]
```diff
@@ -280,13 +465,15 @@ func (h *hardwarePerfEvents) close() {
...
+	fmt.Printf("create hardware perf events")
 	events := &hardwarePerfEvents{
-		cpuCyclesPerfEvents:       []int{},
-		cpuInstructionsPerfEvents: []int{},
-		cacheMissPerfEvents:       []int{},
+		cpuCyclesPerfEvents:       make([]int, 0, numCPU),
+		cpuInstructionsPerfEvents: make([]int, 0, numCPU),
+		cacheMissPerfEvents:       make([]int, 0, numCPU),
 	}
+	fmt.Printf("create hardware perf events: %+v", events)
```
Comment (sthaha) [inline pkg/bpf/exporter.go:474]: log ?

File: pkg/collector/resourceutilization/bpf/process_bpf_collector.go [hunk 7]
```diff
@@ -87,19 +90,28 @@ func UpdateProcessBPFMetrics(bpfExporter bpf.Exporter, processStats map[uint64]*
...
 	}
 	for _, ct := range processesData {
-		comm := C.GoString((*C.char)(unsafe.Pointer(&ct.Comm)))
+		var comm string
+		if _, err := os.Stat("/proc/" + strconv.Itoa(int(ct.Pid))); os.IsNotExist(err) {
+			// skip process that is not running
+			continue
+		}
+		if commBytes, err := os.ReadFile("/proc/" + strconv.Itoa(int(ct.Pid)) + "/comm"); err == nil {
+			comm = strings.TrimSpace(unix.ByteSliceToString(commBytes))
```
Comment (sthaha) [inline pkg/collector/resourceutilization/bpf/process_bpf_collector.go:99]: isn't reading from `/proc` expensive ? instead, can't we pass this data from kernel itself as it used to be? 
Also ,should we re-read the proc every time we process data?

**Review by sthaha — CHANGES_REQUESTED** [review]
Looks good overall, some comments though

File: pkg/bpf/exporter.go (hunk 5 above)
Comment (dave-tucker) [reply pkg/bpf/exporter.go:342]: > Is the purpose of Mutex to lock the entirely of the exporter or only the processMetrics map?

The idea is to lock the entire exporter - so RWMutex won't help.
We want to stop reading new updates from the RingBuffer until after `CollectProcesses` has exited and all the `Total` counters have been reset back to 0.

File: pkg/bpf/exporter.go (hunk 6 above)
Comment (dave-tucker) [reply pkg/bpf/exporter.go:474]: more debug code - removed

File: pkg/collector/resourceutilization/bpf/process_bpf_collector.go (hunk 7 above)
Comment (dave-tucker) [reply pkg/collector/resourceutilization/bpf/process_bpf_collector.go:99]: Re: Getting comm from eBPF... See:
https://github.com/sustainable-computing-io/kepler/issues/1354
https://github.com/sustainable-computing-io/kepler/issues/1640

It's "wrong" for 2 reasons:

1. https://github.com/sustainable-computing-io/kepler/blob/19c6afdccaa602d0a619c251f7ea17d9adb2b2c5/bpf/kepler.bpf.h#L281 `current_comm` is the process that is going OFF CPU, not the process going ON CPU 🐛 
2. Even if we were to use this helper correctly, we probably don't want to be getting the comm on every task switch since it never changes.

Not only that, but `comm` from eBPF is limited to 16 bytes (15 bytes + NUL).
The output from procfs is much nicer, see below.

```
kepler_process_cpu_instructions_total{command="postgres",container_id="fbe5d970ec22cf266028f676f98b5670e1193e33d571eb54f4c9f99f991a2d72",pid="1210084",source="bpf",vm_id=""} 971473
kepler_process_cpu_instructions_total{command="postgres: autovacuum launcher",container_id="fbe5d970ec22cf266028f676f98b5670e1193e33d571eb54f4c9f99f991a2d72",pid="1210248",source="bpf",vm_id=""} 73265
kepler_process_cpu_instructions_total{command="postgres: background writer",container_id="fbe5d970ec22cf266028f676f98b5670e1193e33d571eb54f4c9f99f991a2d72",pid="1210245",source="bpf",vm_id=""} 21080
kepler_process_cpu_instructions_total{command="postgres: walwriter",container_id="fbe5d970ec22cf266028f676f98b5670e1193e33d571eb54f4c9f99f991a2d72",pid="1210247",source="bpf",vm_id=""} 20648
```

We actually have access to the full set of commandline args, but this might be too verbose for use as labels, unless we can find a smarter way of adding them without increasing cardinality.

Re: Performance. It is expensive yes and would benefit from a cache since these values never change between collections.

File: pkg/bpf/exporter.go (hunk 5 above)
Comment (sthaha) [reply pkg/bpf/exporter.go:341]: Thanks for the clarification.  Would you mind moving the mutex to the top of struct ? 
You can find similar patterns in go . The idea is to keep mutex above what it protects.

File: pkg/bpf/exporter.go [hunk 8]
```diff
@@ -192,38 +193,223 @@ func (e *exporter) Detach() {
...
-		)
-		total += count
-		if errors.Is(err, ebpf.ErrKeyNotExist) {
-			break
+		var record *ringbuf.Record
+
+		select {
+		case <-stopChan:
+			if err := rd.Close(); err != nil {
+				return fmt.Errorf("closing ring buffer reader: %v", err)
```
Comment (sthaha) [inline pkg/bpf/exporter.go:218]:
```suggestion
				return fmt.Errorf("closing ring buffer reader: %w", err)
```

File: pkg/bpf/exporter.go [hunk 9]
```diff
@@ -271,6 +457,7 @@ type hardwarePerfEvents struct {
 }
 
 func (h *hardwarePerfEvents) close() {
+	fmt.Printf("close perf events: %+v", h)
```
Comment (sthaha) [inline pkg/bpf/exporter.go:460]: klog?

File: pkg/bpf/exporter.go [hunk 10]
```diff
@@ -280,13 +467,14 @@ func (h *hardwarePerfEvents) close() {
 // and updates the corresponding eBPF maps.
 func createHardwarePerfEvents(cpuInstructionsMap, cpuCyclesMap, cacheMissMap *ebpf.Map, numCPU int) (*hardwarePerfEvents, error) {
 	var err error
+	fmt.Printf("create hardware perf events")
```
Comment (sthaha) [inline pkg/bpf/exporter.go:470]: klog

File: pkg/comm/resolve_comm.go [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+
+	comm, err := readCommandFromProcFs(pid)
+	if err != nil {
+		if os.IsNotExist(err) {
+			// skip process that is not running
+			r.cacheNotExist[pid] = true
+			return unknownComm, fmt.Errorf("process not running: %w", err)
+		}
+	}
+	r.cacheExist[pid] = comm
```
Comment (sthaha) [inline pkg/comm/resolve_comm.go:43]: when do we invalidate both caches?

File: pkg/comm/resolve_comm.go [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+	}
+
+	comm, err := readCommandFromProcFs(pid)
+	if err != nil {
+		if os.IsNotExist(err) {
+			// skip process that is not running
+			r.cacheNotExist[pid] = true
+			return unknownComm, fmt.Errorf("process not running: %w", err)
+		}
+	}
```
Comment (sthaha) [inline pkg/comm/resolve_comm.go:42]:
```suggestion
	if err != nil && os.IsNotExist(err) {
			// skip process that is not running
			r.cacheNotExist[pid] = true
			return unknownComm, fmt.Errorf("process not running: %w", err)
		}
	}
```

File: pkg/bpf/exporter.go [hunk 13]
```diff
@@ -192,38 +193,223 @@ func (e *exporter) Detach() {
...
-	// Get the max number of entries in the map
-	maxEntries := e.bpfObjects.Processes.MaxEntries()
-	total := 0
-	deleteKeys := make([]uint32, maxEntries)
-	deleteValues := make([]ProcessMetrics, maxEntries)
-	var cursor ebpf.MapBatchCursor
+func (e *exporter) Start(stopChan <-chan struct{}) error {
+	rd, err := ringbuf.NewReader(e.bpfObjects.Rb)
+	if err != nil {
+		return fmt.Errorf("failed to create ring buffer reader: %v", err)
```
Comment (sthaha) [inline pkg/bpf/exporter.go:208]:
```suggestion
		return fmt.Errorf("failed to create ring buffer reader: %w", err)
```

**Review by sthaha — COMMENTED** [review]
Looks good overall. I think we should address cache invalidation of `comm`.

File: pkg/bpf/exporter.go (hunk 5 above)
Comment (dave-tucker) [reply pkg/bpf/exporter.go:348]: I've kept it where it was and added code comments. It is protecting processMetrics but also freedPIDs.

File: pkg/collector/resourceutilization/bpf/process_bpf_collector.go (hunk 7 above)
Comment (dave-tucker) [reply pkg/collector/resourceutilization/bpf/process_bpf_collector.go]: Cache added.

File: pkg/comm/resolve_comm.go (hunk 11 above)
Comment (dave-tucker) [reply pkg/comm/resolve_comm.go:53]: fixed. we now listen for `sched_process_free` events and evict stale PIDs from both caches every sample period.

**Comment by dave-tucker** [comment]
@sthaha This is now ready for another look. Caching of `comm` from procfs is now solved with another eBPF probe that tells us when PIDs have been freed. I've also marked this as fixing all the issues relating to the process `comm` since these are now accurate:

For example:
```console
$ curl -s localhost:8888/metrics | grep grafana-server | grep cpu_cycles           
kepler_process_cpu_cycles_total{command="grafana-server",container_id="f3e1cb741182257a9fcfb8e71ccf9ef14e044556f6225900323dce7a10cb2ac0",pid="2025102",source="bpf",vm_id=""} 1.593157176e+09
```

``` console
$  sudo ps aux | grep grafana-server
nobody   2025102  0.4  0.1 797072 101700 ?       Ssl  Jul11 117:49 grafana-server --homepath=/usr/share/grafana --config=/etc/grafana/grafana.ini --packaging=docker cfg:default.log.mode=console cfg:default.paths.data=/var/lib/grafana cfg:default.paths.logs=/var/log/grafana cfg:default.paths.plugins=/var/lib/grafana/plugins cfg:default.paths.provisioning=/etc/grafana/provisioning
```

Once this merges I'll open an issue to discuss this further as I think we should probably move this type of data into logs vs adding it as a label - but that would require changes to the operator further down the line.

File: bpf/kepler.bpf.c [hunk 14]
```diff
@@ -3,6 +3,168 @@
 
 #include "kepler.bpf.h"
 
+// Ring buffer sizing
+// 256kB is sufficient to store around 1000 events/sec for 5 seconds
```
Comment (sthaha) [inline bpf/kepler.bpf.c:7]: ❤️

File: pkg/bpf/exporter.go [hunk 15]
```diff
@@ -35,23 +38,38 @@ import (
...
 	pageWriteLink   link.Link
 	pageReadLink    link.Link
+	processFreeLink link.Link
 
 	perfEvents *hardwarePerfEvents
 
 	enabledHardwareCounters sets.Set[string]
 	enabledSoftwareCounters sets.Set[string]
+
+	// Locks processMetrics and freedPIDs.
```
Comment (sthaha) [inline pkg/bpf/exporter.go:54]: ❤️

**Review by sthaha — APPROVED** [review]
LGTM! Thanks a lot for incorporating the comments 🙇

**Comment by rootfs** [comment]
I did a quick test to evaluate the eBPF latency reduction with regard to 0.7.8 (which has both tracking and sampling modes).

# TL;DR 
In this current implementation, using ringbuffer reduces eBPF tracking mode latency by 4.5x, but still a long shot from sampling mode performance. 

# Result 

| Latest code with this PR, sample rate 1000| 0.7.8 code, tracking (no sampling)|0.7.8 code, sampling rate 1000|
|---|---|---|
| 484 ns | 2043 ns | 78 ns | 

# Test
[asciicast](https://asciinema.org/a/670546)

# Test script
The test script is [here](https://gist.github.com/rootfs/ae1e50cc69bd5d07dc2e0f5df3ea3869)

**Comment by dave-tucker** [comment]
> In this current implementation, using ringbuffer reduces eBPF tracking mode latency by 4.5x, but still a long shot from sampling mode performance.

It's always a compromise between accuracy and performance.
Look at it this way:

| Case | Duration | Accuracy |
|------|----------|----------|
| latest | 484ns | 100% |
| v0.7.8 | 2043ns | 100% |
| v0.7.8 Sample Rate 1000 | 78ns | 0.1% |

The ringbuf case to me seems to be the best of both worlds. Acceptable performance with the best accuracy.
